docs: update go-pull.md (#2844)
grafana/pyroscope: docs/sources/configure-client/grafana-agent/go_pull.md

---

> title: "Go (pull mode)"
> menuTitle: "Go (pull mode)"

Reviewer: ```suggestion
title: "Set up Go profiling in pull mode
menuTitle: Set up Go profiling in pull mode
```

> description: "Set up Go profiling in pull mode"

Reviewer: Quotes aren't necessary in the description, title, and menuTitle unless there are special characters (like parentheses). 
```suggestion
description: Learn how to set up Go profiling in pull mode.
```

weight: 10
---

> # Go (pull mode)

Reviewer: I really like the description you used as a page title and heading. What do you think about using it as the page title and heading? 
```suggestion
# Set up Go profiling in pull mode
```

In pull mode, the Grafana Agent periodically retrieves profiles from Golang applications, specifically targeting the
`/debug/pprof/*` endpoints.

> ## Set up Go profiling in pull mode

Reviewer: Because this section does list what a user needs to have before using the procedures on this page, consider using Before you begin (which is the standard heading we use in other procedures). 
```suggestion
## Before you begin
```

> ## Set up Go profiling in pull mode

Reviewer: Really good changes! Love the steps you've added and the matching headings. 

We probably don't need a heading here since the content below is really part of the introduction. However, are there any prerequisites that people need to have before continguing into the procedures? If so, consider adding a "Before you begin" section that lists those requirements. As an example, refer to https://grafana.com/docs/tempo/latest/setup/tanka/
```suggestion
```

Author: The only prerequisite is to have grafana agent, and I put it not as prerequisite so not needed

To set up Golang profiling in pull mode, you need to:

1. Expose pprof endpoints
2. Install Grafana Agent
3. Prepare Grafana Agent configuration file
4. Start Grafana Agent

### Expose pprof endpoints

Ensure your Golang application exposes pprof endpoints.

1. Get `godeltaprof` package

    ```bash
    go get github.com/grafana/pyroscope-go/godeltaprof@latest
    ```

2. Import `net/http/pprof` and `godeltaprof/http/pprof` packages at the start of your application.

    ```go
    import _ "net/http/pprof"
    import _ "github.com/grafana/pyroscope-go/godeltaprof/http/pprof"
    ```

### Install Grafana Agent

> [//]: # (TODO&#40;korniltsev&#41; What should go here?)

Reviewer: Are there specific things that someone has to do to Agent configuration to use Go in pull mode? If so, those things should be added here. 

At a minimum, we should link to the Agent installation doc. If we have more time, ideally we should include a config example for Agent for Pyroscope-specific config options. The latter would make the doc more usable. 

Here's another example. In the [Set up a test app doc in Tempo](https://grafana.com/docs/tempo/latest/setup/set-up-test-app/), we use an example Agent config as part of the procedure. (We have a PR that replaces that example with Flow instead of Static mode.) 

The minimum viable doc option here is a link agent installation procedures. We can always iterate and add more specifics. If this is the option you choose, you should make sure to create a doc issue so the update to add Agent config doesn't get lost.

Reviewer: Ha! And underneath you have the configuration file! Never mind! :)

> [flow mode](/docs/agent/latest/flow/setup/install/)

Reviewer: Please feel free to reword this text as you see fit so it matches 
```suggestion
This procedure uses Grafana Agent to send data to Pyroscope using an example River configuration file for the Grafana Agent in Flow mode. 
Make sure you have [Grafana Agent in Flow mode](/docs/agent/latest/flow/setup/install/) installed. 
```

> ### Prepare Grafana Agent Flow configuration file

Reviewer: ❤️

In the Grafana Agent Flow configuration file, you need to add at least two blocks: `pyroscope.write`

> and `pyroscope.scrape`.

Reviewer: Should we mention the location and name of the configuration file? Or a default location or file name? (Understanding that this may vary depending upon the operating system.)

Author: location and file name does not matter, it can be /foo/bar239.png

            forward_to = [pyroscope.write.write_job_name.receiver]
    
            profiling_config {
                    profile.process_cpu {
                            enabled = true
                    }
    
                    profile.godeltaprof_memory {
                            enabled = true
                    }
    
                    profile.memory { // disable memory, use godeltaprof_memory instead
                            enabled = false
                    }
    
                    profile.godeltaprof_mutex {
                            enabled = true
                    }
    
                    profile.mutex { // disable mutex, use godeltaprof_mutex instead
                            enabled = false
                    }
    
                    profile.godeltaprof_block {
                            enabled = true
                    }
    
                    profile.block { // disable block, use godeltaprof_block instead
                            enabled = false
                    }
    
                    profile.goroutine {
                            enabled = true
                    }
            }
    }

    ```

> 

Reviewer: ```suggestion
3. Save the changes to the file. 
```

### Start Grafana Agent Flow

> 1. Start a local pyroscope instance for testing purposes

Reviewer: ```suggestion
1. Start a local Pyroscope instance for testing purposes
```

    ```bash
    docker run -p 4040:4040 grafana/pyroscope 
    ```
2. Start Grafana Agent
    ```bash
    grafana-agent-flow run conifguration.river
    ```

> 3. Go to http://localhost:4040 and you should see profiles there.

Reviewer: ```suggestion
3. Open a browser to http://localhost:4040. The page should list profiles. 
```

            rule {
                    action        = "replace"
                    source_labels = ["__meta_kubernetes_pod_name"]
                    target_label  = "pod"
            }
    
            rule {
                    action        = "replace"
                    source_labels = ["__meta_kubernetes_node_name"]
                    target_label  = "node"
            }
    
            rule {
                    action        = "replace"
                    source_labels = ["__meta_kubernetes_pod_container_name"]
                    target_label  = "container"
            }
    
            rule {
                    action        = "replace"
                    regex         = "(.*)@(.*)"
                    replacement   = "${1}/${2}"
                    separator     = "@"
                    source_labels = ["__meta_kubernetes_namespace", "__meta_kubernetes_pod_container_name"]
                    target_label  = "service_name"
            }
            
            rule { 
                    action        = "keep"
                    regex         = "(ns1/.*)|(ns2/container-.*0)"
                    source_labels = ["service_name"]
            }
    }
    ```

> 3. And then use `discovery.relabel.specific_pods.targets` as a target for `pyroscope.scrape` block.

Reviewer: ```suggestion
3. Use `discovery.relabel.specific_pods.targets` as a target for `pyroscope.scrape` block.
```

    ```river
        pyroscope.scrape "scrape_job_name" {
                targets    = discovery.relabel.specific_pods.output
                ...
        }
    ```

### Exposing pprof endpoints

> If you don't use `http.DefaultServeMux` you can register `/debug/pprof/*` handlers to your own `http.ServeMux`

Reviewer: ```suggestion
If you don't use `http.DefaultServeMux`, you can register `/debug/pprof/*` handlers to your own `http.ServeMux`
```

> ### Exposing pprof endpoints

Reviewer 2: Should this be at the end or the beginning? seems like we should have people expose their pprof endpoints and _then_ do the rest of the steps to scrape from them

Reviewer 2: either way, I think we should put this next to, if not in, the "exposing pprof endpoints" section. I suspect someone who was looking for this would expect to find it near or in that section

Author: lets remove it alltogether. I dont want this to be next to the  "exposing pprof endpoints" as it complicates it

```go
var mux *http.ServeMux
mux.Handle("/debug/pprof/", http.DefaultServeMux)
```

> Or if you use gorilla/mux:

Reviewer: ```suggestion
Or, if you use gorilla/mux:
```

```go
var router *mux.Router
router.PathPrefix("/debug/pprof").Handler(http.DefaultServeMux)
```

> 

Reviewer: What are the next steps for someone? is there a page that we can link them to for the next step in what they need to do? (Or maybe we need to write one?)

Author: not sure I understand your question. The last steps are: see profiles in browser, optimize your application, repeat

Reviewer: Sometimes there is a series of tasks you need to do. Those tasks may be on multiple pages. The "next step" can be a link to the procedure the user needs to perform next. if that isn't the case here, then we don't need a link to another page. 

However, if there are things a user should repeat to refine the results, we might want to mention that.
